planner/core/task.go

```diff
@@ -581,22 +587,31 @@ func needDecimalConvert(tp *types.FieldType, rtp *types.FieldType) bool {
 }
 
 func negotiateCommonType(lType, rType *types.FieldType) (*types.FieldType, bool, bool) {
-	lExtend := 0
-	rExtend := 0
-	cDec := rType.Decimal
-	if lType.Decimal < rType.Decimal {
-		lExtend = rType.Decimal - lType.Decimal
-	} else if lType.Decimal > rType.Decimal {
-		rExtend = lType.Decimal - rType.Decimal
-		cDec = lType.Decimal
-	}
-	lLen, rLen := lType.Flen+lExtend, rType.Flen+rExtend
-	cLen := mathutil.Max(lLen, rLen)
-	cLen = mathutil.Min(65, cLen)
-	cType := types.NewFieldType(mysql.TypeNewDecimal)
-	cType.Decimal = cDec
-	cType.Flen = cLen
-	return cType, needDecimalConvert(lType, cType), needDecimalConvert(rType, cType)
+	commonType := types.AggFieldType([]*types.FieldType{lType, rType})
+	if commonType.Tp == mysql.TypeNewDecimal {
+		lExtend := 0
+		rExtend := 0
+		cDec := rType.Decimal
+		if lType.Decimal < rType.Decimal {
+			lExtend = rType.Decimal - lType.Decimal
+		} else if lType.Decimal > rType.Decimal {
+			rExtend = lType.Decimal - rType.Decimal
+			cDec = lType.Decimal
+		}
+		lLen, rLen := lType.Flen+lExtend, rType.Flen+rExtend
+		cLen := mathutil.Max(lLen, rLen)
+		cLen = mathutil.Min(65, cLen)
+		commonType.Decimal = cDec
+		commonType.Flen = cLen
+	} else if needConvert(lType, commonType) || needConvert(rType, commonType) {
+		if mysql.IsIntegerType(commonType.Tp) {
+			// If the target type is int, both TiFlash and Mysql only support cast to Int64
+			// so we need to promote the type to Int64
+			commonType.Tp = mysql.TypeLonglong
+			commonType.Flen = mysql.MaxIntWidth
+		}
```

Comment:
what if int(x) = decimal(n,m)?

Reply:
TiDB guarantees that two type are in the same catalog, so it should not happen here.